### ISSUE TITLE
SQLFORM.grid raises KeyError for an invalid "order" parameter

### ISSUE THREAD

**Issue by dartg** [web2py/web2py#1422]
SQLFORM.grid uses the `order` parameter to know which column needs to be sorted. Something like:

`http://.../a/c/f?keywords=&order=mytable.myfield&_signature=...`

The problem is, changing `order=mytable.myfield` to `order=mytable.myfield1` (or any random string which doesn't satisfy the `table.field` structure) raises a KeyError. 

```
File "C:\web2py\gluon\sqlhtml.py", line 2545, in grid
    sort_field = db[otablename][ofieldname]
```

It would be better if that line could be wrapped inside a `try/except KeyError` and just ignore the `order` parameter. 

As a workaround, until this is fixed, the following code could be inserted before calling `SQLFORM.grid`, to ignore the order if an invalid value is provided.

```
if request.vars.order is not None:
    try:
        table, field = request.vars.order.split(".")
    except ValueError:
        request.vars.order = None
    else:
        if table.startswith("~"):
            table = table[1:]
        if table not in db or field not in db.mytable.fields:
            request.vars.order = None
```

**Comment by niphlod** [comment]
ahem. why something that is required should be discarded if not valid ?

**Comment by dartg** [comment]
I'm afraid I don't understand. What do you mean?

**Comment by niphlod** [comment]
@dartg : the order parameter is required for the grid to work (as many other). 
Why should the grid be able to ingest something that doesn't translate to a proper ordering ? if the order parameter is there, it needs to be a valid table.column identifier.

**Comment by dartg** [comment]
I think the problem with this is that SQLFORM.grid is an API which uses internally the order parameter. The user implementing that API has no idea what `request.vars.order` is, so why the should be responsible for handling an error of that parameter? It's like SQLFORM.grid says to the user "I'll take care about columns ordering" but, when an error occurs, it doesn't take care anymore. The user can't guess how the API implements ordering internally until he reads the source (since this behavior isn't documented either).

In short, the order parameter is implemented by SQLFORM.grid, not by the user, so it's the former's responsibility when an error occurs.

I talk as a web2py user. What do you think?

**Comment by niphlod** [comment]
That this is not the spirit of "we're all consenting adults". I'd refrain
from typechecking at each and every request every option and argument
passed. When a user inserts a grid in a page, the grid takes control and
the user shouldn't fiddle without proper knowledge. Order,Page etc are all
required to comply for the grid displaying something. Passing over would
mask an otherwise perfectly valid exception

Il 02 ott 2016 9:39 PM, "dartg" <notifications@github.com> ha scritto:

**Comment by dartg** [comment]
I'd also refrain from typechecking, but I think web2py should notice the user that the order parameter is not valid. A `KeyError` may be raised by other parts of the SQLFORM.grid code, so the user can't rely on that. I still think no one will care if "order" failed (will just ignore it), but with the current trunk the user is not able to know if order failed until it implements some code like the one above, which after all is typecheking before calling SQLFORM.grid.
